File: 01-Listas-Enlazadas/C/listaenlazada.c
#include "nodo.c"
#include <stdio.h>
#include <stdlib.h>

// lista enlazada simple creada para manejar datos genericos
// desarrollada de esta forma solo para fines didacticos
// es mas conveniente desarrorla directamente para manejar el tipo de dato especifico que se desea usar
typedef struct lista {
    nodo *inicio;
    nodo *fin;
    int longitud;
}lista;

// crear una nueva lista enlazada
// retorna un puntero a la lista
lista *crearlista() {
    lista *list = malloc(sizeof(lista)); // reserva memoria en heap para la lista
    list->longitud = 0;
    list->inicio = NULL;
    list->fin = NULL;
    return list;
}

// agregar un elemento en el indice pos
// si pos es -1 agregar al final
void agregar(lista *list, void *dato, int pos) {
    if (pos < -1 || pos > list->longitud) return;

    if (list->longitud == 0) {
        nodo *new = newnodo(dato);
        list->inicio = new;
        list->fin = new;
        list->longitud = 1;
    }
    else if (list->longitud == 1 && (pos == 1 || pos == -1)) {
        nodo *new = newnodo(dato);
        list->fin = new;
        list->inicio->siguiente = list->fin;
        list->longitud++;
    }
    else if (pos == 0) {
        nodo *new = newnodo(dato);
        new->siguiente = list->inicio;
        if (list->longitud == 1) list->fin = list->inicio;
        list->inicio = new;
        list->longitud++;
    }
    else if (pos == list->longitud || pos == -1) {
        nodo *new = newnodo(dato);
        list->fin->siguiente = new;
        list->fin = new;
        list->longitud++;
    }
    else {
        nodo *tmp = list->inicio;
        nodo *new = newnodo(dato);
        for (int i = 0; i < pos - 1; i++) {
            tmp = tmp->siguiente;
        }
        new->siguiente = tmp->siguiente;
        tmp->siguiente = new;
        list->longitud++;
    }
}

// elminar el elmento en el indice pos
// si pos es -1 eliminar el ultimo elemento
void eliminar(lista *list, int pos) {
    if (pos < -1 || pos >= list->longitud) return;
    if (list->longitud == 0) return;

    if (list->longitud == 1) {
        free(list->inicio); // liberamos la memoria del nodo
        list->inicio = NULL;
        list->fin = NULL;
        list->longitud = 0;
    }
    else if (pos == 0) {
        nodo *tmp = list->inicio->siguiente;
        free(list->inicio); // liberamos la memoria del nodo
        list->inicio = tmp;
        list->longitud--;
    }
    else if (list->longitud == 2 && (pos == 1 || pos == -1)) {
        free(list->fin); // liberamos la memoria del nodo
        list->inicio->siguiente = NULL;
        list->fin = list->inicio;
        list->longitud--;
    }
    else {
        int n = pos == -1 ? list->longitud - 1 : pos;
        nodo *actual = list->inicio;
        for (int i = 0; i < n - 1; i++) {
            actual = actual->siguiente;
        }
        nodo *sig = actual->siguiente->siguiente;
        free(actual->siguiente); // liberamos la memoria del nodo
        actual->siguiente = sig;
        if (n == list->longitud - 1) list->fin = actual;
        list->longitud--;
    }

}

// obtener el valor en el indice dado
// si el indice es -1 obtener el ultimo elemento
// solo funciona si el tipo de dato es int
void *obtener(lista *list, int indice) {
    if (indice < -1 || indice >= list->longitud) return NULL;
    if (indice == 0) return list->inicio->dato;
    if (indice == -1) return list->fin->dato;
        nodo *tmp = list->inicio;
        for (int i = 0; i < indice; i++) {
            tmp = tmp->siguiente;
        }
        return tmp->dato;
}

// buscar el indice del dato dado
// si no se encuentra retorna -1
// solo funciona si el tipo de dato es int
int buscar(lista *list, int dato) {
    if (dato == *(int*)(list->inicio->dato)) return 0;
    if (dato == *(int*)(list->fin->dato)) return list->longitud - 1;
    nodo *current = list->inicio;
    int index = 0;
    while (current != NULL) {
        if (*(int*)(current->dato) == dato) {
            return index;
        }
        current = current->siguiente;
        index++;
    }
    return -1;
}


// imprimir la lista por consola
// solo funciona si el tipo de dato es int
void printlist(lista *list) {
    if (list->longitud == 0) {
        printf("empty list\n");
        return;
    }
    int len = list->longitud;
    nodo *current = list->inicio;
    for (int i = 0; i < len; i++) {
        if (i != len - 1)printf("%d -> ", *(int*)(current->dato));
        else printf("%d", *(int*)(current->dato));
        current = current->siguiente;
    }
    printf("\n");
}

File: 01-Listas-Enlazadas/Java/ListaEnlazada.java
public class ListaEnlazada<T> {
    
    // nodos
    Nodo<T> primero;
    Nodo<T> ultimo;
    int longitud;

    public ListaEnlazada() {
        primero = ultimo = null;
        longitud = 0;
    }

    // agregar un elemento al final
    public void agg(T dato) {
        Nodo<T> nodo = new Nodo<>(dato);
        if (estaVacio()) primero = ultimo = nodo;
        else if (longitud == 1) {
            ultimo = nodo;
            primero.setSig(ultimo);
        }
        else {
            ultimo.setSig(nodo);
            ultimo = nodo;
        }
        longitud++;
    }

    // agregar un elemento en una posicion dada
    public void agg(T dato, int pos) {
        if (pos < 0 || pos > longitud) throw new IndexOutOfBoundsException();
        if (estaVacio() || longitud == 1 || pos == longitud) agg(dato);
        else if (pos == 0) {
            Nodo<T> nodo = new Nodo<>(dato);
            nodo.setSig(primero);
            primero = nodo;
            longitud++;
        }
        else {
            Nodo<T> nodo = new Nodo<>(dato);
            Nodo<T> aux = primero;
            for (int i = 0; i < pos - 1; i++) {
                aux = aux.getSig();
            }
            nodo.setSig(aux.getSig());
            aux.setSig(nodo);
            longitud++;
        }
    }

    // eliminar el ultimo elemento
    public void elim() {
        if (estaVacio()) throw new RuntimeException("Lista vacia");
        if (longitud == 1) primero = ultimo = null;
        else if (longitud == 2) {
            ultimo = primero;
            primero.setSig(null);
        }
        else {
            Nodo<T> aux = primero;
            for (int i = 1; i < longitud - 1; i++) {
                aux = aux.getSig();
            }
            aux.setSig(null);
            ultimo = aux;
        }
        longitud--;
    }

    // eliminar el elemento en indice dado
    public void elim(int pos) {
        if (estaVacio()) throw new RuntimeException("Lista vacia");
        if  (pos < 0 || pos > longitud) throw new IndexOutOfBoundsException();
        if (pos == longitud - 1) elim();
        else if (pos == 0) {
            primero = primero.getSig();
            longitud--;
        }
        else {
            Nodo<T> aux = primero;
            for (int i = 0; i < pos - 1; i++) {
                aux = aux.getSig();
            }
            aux.setSig(aux.getSig().getSig());
            longitud--;
        }
    }

    // buscar dato, retorna el indice de su posicion o -1 si no lo encuentra
    public int buscar(T dato) {
        if (estaVacio()) throw new RuntimeException("Lista vacia");
        if (primero.getDato().toString().equals(dato.toString())) return 0;
        else if (ultimo.getDato().toString().equals(dato.toString())) return longitud - 1;
        else {
            Nodo<T> aux = primero;
            int pos = 0;
            while (aux != null) {
                if (aux.getDato().toString().equals(dato.toString())) return pos;
                aux = aux.getSig();
                pos++;
            }
            return -1;
        }
    }

    // obtener dato en indice dado
    public T obtener(int pos) {
        if (estaVacio()) throw new RuntimeException("Lista vacia");
        if (pos < 0 || pos > longitud) throw new IndexOutOfBoundsException();
        if (pos == 0)  return primero.getDato();
        else if (pos == longitud - 1) return ultimo.getDato();
        else {
            Nodo<T> aux = primero;
            for (int i = 0; i < pos; i++) {
                aux = aux.getSig();
            }
            return aux.getDato();
        }
    }

    public boolean estaVacio() {
        return longitud == 0;
    }


    @Override
    public String toString() {
        Nodo<T> nodo = primero;
        String rsp = "";
        while (nodo != null) {
            if (nodo.getSig() != null)rsp += nodo.getDato().toString() + " -> ";
            else rsp += nodo.getDato().toString();
            nodo = nodo.getSig();
        }
        return rsp;
    }
}
